bigtable/metrics.go

```diff
@@ -0,0 +1,479 @@
+	counters      map[string]metric.Int64Counter     // Key is metric name e.g. retry_count
+}
+
+// Generates unique client ID in the format go-<random UUID>@<>hostname
+func generateClientUID() (string, error) {
+	hostname := "localhost"
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", err
+	}
+	return "go-" + uuid.NewString() + "@" + hostname, nil
+}
+
+type metricsConfigInternal struct {
+	exporter otelmetric.Exporter
+
+	project string
+
+	// Flag to indicate whether built-in metrics should be recorded and exported to GCM
+	builtInEnabled bool
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	// Contains one entry per meter provider
+	instruments []builtInMetricInstruments
+}
+
+func newMetricsConfigInternal(ctx context.Context, project, instance string, userProvidedConfig *metricsConfig, opts ...option.ClientOption) (*metricsConfigInternal, error) {
+	clientUID, err := generateClientUID()
+	if err != nil {
+		log.Printf("built-in metrics: generateClientUID failed: %v. Using empty string in the %v metric atteribute", err, metricLabelKeyClientUID)
+	}
+
+	// Create default meter provider
+	internalMetricsConfig := &metricsConfigInternal{
+		project: project,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+		},
+		instruments:    []builtInMetricInstruments{},
+		builtInEnabled: builtInEnabledDefault,
+	}
+
+	defaultExporter, err := newMonitoringExporter(ctx, internalMetricsConfig.project, opts...)
+	if err != nil {
+		return nil, err
+	}
+	internalMetricsConfig.exporter = defaultExporter
+
+	// Create default meter provider
+	defaultMp := otelmetric.NewMeterProvider(
+		otelmetric.WithReader(
+			otelmetric.NewPeriodicReader(
+				defaultExporter,
+				otelmetric.WithInterval(defaultSamplePeriod),
+			),
+		),
+	)
+
+	userProvidedMeterProviders := []*otelmetric.MeterProvider{}
+	if addr := os.Getenv("BIGTABLE_EMULATOR_HOST"); addr != "" {
+		// Do not emit metrics when emulator is being used
+		internalMetricsConfig.builtInEnabled = false
+	} else if userProvidedConfig != nil {
+		internalMetricsConfig.builtInEnabled = userProvidedConfig.builtInEnabled
+		userProvidedMeterProviders = userProvidedConfig.meterProviders
+	}
+
+	if !internalMetricsConfig.builtInEnabled {
+		return internalMetricsConfig, nil
+	}
+
+	// Create instruments on all meter providers
+	allMeterProviders := append(userProvidedMeterProviders, defaultMp)
+	for _, mp := range allMeterProviders {
+		builtInMetricInstruments := builtInMetricInstruments{
+			distributions: make(map[string]metric.Float64Histogram),
+			counters:      make(map[string]metric.Int64Counter),
+		}
```

Comment:
I'm a bit confused why you need the indirection via the hashmaps? Why not have the instruments be struct fields?

Reply:
It was so that I could create instruments in a loop. Removed the hashmap now.